<a href="https://colab.research.google.com/github/gapself/machine-learning-projects/blob/main/data-analysis/diagnosing_diabetes_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# EDA: Diagnosing Diabetes

https://www.kaggle.com/datasets/uciml/pima-indians-diabetes-database

## Initial Inspection

1. First, familiarize yourself with the dataset.
   Look at each of the nine columns in the documentation.
   
   What do you expect each data type to be?

Expected data type for each column:

- `Pregnancies`: 
- `Glucose`: 
- `BloodPressure`: 
- `SkinThickness`: 
- `Insulin`: 
- `BMI`: 
- `DiabetesPedigreeFunction`: 
- `Age`: 
- `Outcome`: 

2. Next, let's load in the diabetes data to start exploring.

   Load the data in a variable called `diabetes_data` and print the first few rows.
   
   **Note**: The data is stored in a file called `diabetes.csv`.

In [52]:
import pandas as pd
import numpy as np

# load in data
diabetes_data = pd.read_csv('diabetes.csv')
print(diabetes_data.columns)

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')


3. How many columns (features) does the data contain?

In [10]:
# print number of columns
print(len(diabetes_data.columns))

9


4. How many rows (observations) does the data contain?

In [13]:
# print number of rows
print(diabetes_data.shape[0])

768


## Further Inspection

5. Let's inspect `diabetes_data` further.

   Do any of the columns in the data contain null (missing) values?

In [16]:
# find whether columns contain null values
print(diabetes_data.isnull().any())

Pregnancies                 False
Glucose                     False
BloodPressure               False
SkinThickness               False
Insulin                     False
BMI                         False
DiabetesPedigreeFunction    False
Age                         False
Outcome                     False
dtype: bool


6. If you answered no to the question above, not so fast!

   While it's technically true that none of the columns contain null values, that doesn't necessarily mean that the data isn't missing any values.
   
   When exploring data, you should always question your assumptions and try to dig deeper.
   
   To investigate further, calculate summary statistics on `diabetes_data` using the `.describe()` method.

In [17]:
# perform summary statistics
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000                  0.243750   24.000000  
50%     32.000000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000

7. Looking at the summary statistics, do you notice anything odd about the following columns?

   - `Glucose`
   - `BloodPressure`
   - `SkinThickness`
   - `Insulin`
   - `BMI`

**Your response to question 7**:


>  If you take a look at the minimum values for these five columns, you'll notice that they are all 0. How can Blood Pressure or BMI be 0? That makes no sense! These values also seem to be way off from their respective medians and means, another indicator that something is off. One way to interpret this is that there are missing values in the data.



8. Do you spot any other outliers in the data?

**Your response to question 8**: 

>   eg. max pregnancies: 17 <br> max insulin: 846



9. Let's see if we can get a more accurate view of the missing values in the data.

   Replace the instances of `0` with `NaN` in the five columns mentioned:
   
   ```py
   diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']]
   ```

In [51]:
# replace instances of 0 with NaN
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)

10. Next, check for missing (null) values in all of the columns just like you did in Step 5.

    Now how many missing values are there?

In [21]:
# find whether columns contain null values after replacements are made
print(diabetes_data.isnull().any())

Pregnancies                 False
Glucose                      True
BloodPressure                True
SkinThickness                True
Insulin                      True
BMI                          True
DiabetesPedigreeFunction    False
Age                         False
Outcome                     False
dtype: bool


In [25]:
print(diabetes_data.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


11. Let's take a closer look at these rows to get a better idea of _why_ some data might be missing.

    Print out all the rows that contain missing (null) values.

In [27]:
# print rows with missing values
print(diabetes_data[diabetes_data.isnull().any(axis=1)])

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
..           ...      ...            ...            ...      ...   ...   
761            9    170.0           74.0           31.0      NaN  44.0   
762            9     89.0           62.0            NaN      NaN  22.5   
764            2    122.0           70.0           27.0      NaN  36.8   
766            1    126.0           60.0            NaN      NaN  30.1   
767            1     93.0           70.0           31.0      NaN  30.4   

     DiabetesPedigreeFunction  Age Outcome  
0                       0.627   50       1  
1                    

12. Go through the rows with missing data. Do you notice any patterns or overlaps between the missing data?

**Your response to question 12**:


> One thing you might notice is that most rows with missing data have missing values in more than one column. In fact, every single row with at least one missing value also has a missing value in the Insulin column. This is a clue as to why the data is missing! If patients did not have their insulin measured, why might they also not have had these other measurements taken? Depending on how much data is missing, you might choose to remove specific rows or impute the missing values somehow.



13. Next, take a closer look at the data types of each column in `diabetes_data`.

    Does the result match what you would expect?

In [29]:
# print data types using .info() method
print(diabetes_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(6), int64(2), object(1)
memory usage: 54.1+ KB
None


14. To figure out why the `Outcome` column is of type `object` (string) instead of type `int64`, print out the unique values in the `Outcome` column.

In [30]:
# print unique values of Outcome column
print(diabetes_data['Outcome'].unique())

['1' '0' 'O']


15. How might you resolve this issue?

**Your response to question 15**:


> replace letter with number 0, and then convert column to int




## Next Steps:

16. Congratulations! In this project, you saw how EDA can help with the initial data inspection and cleaning process. This is an important step as it helps to keep your datasets clean and reliable.

    Here are some ways you might extend this project if you'd like:
    - Use `.value_counts()` to more fully explore the values in each column.
    - Investigate other outliers in the data that may be easily overlooked.
    - Instead of changing the `0` values in the five columns to `NaN`, try replacing the values with the median or mean of each column.

In [32]:
diabetes_data.value_counts()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  BMI   DiabetesPedigreeFunction  Age  Outcome
0            74.0     52.0           10.0           36.0     27.8  0.269                     22   0          1
4            117.0    64.0           27.0           120.0    33.2  0.230                     24   0          1
             111.0    72.0           47.0           207.0    37.1  1.390                     56   1          1
             110.0    76.0           20.0           100.0    28.4  0.118                     27   0          1
             109.0    64.0           44.0           99.0     34.8  0.905                     26   1          1
                                                                                                            ..
1            131.0    64.0           14.0           415.0    23.7  0.389                     21   0          1
             130.0    70.0           13.0           105.0    25.9  0.472                     22   0          1
      

In [97]:
diabetes_data = pd.read_csv('diabetes.csv')

In [98]:
diabetes_data.head(100)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             6      148             72             35        0  33.6   
1             1       85             66             29        0  26.6   
2             8      183             64              0        0  23.3   
3             1       89             66             23       94  28.1   
4             0      137             40             35      168  43.1   
..          ...      ...            ...            ...      ...   ...   
95            6      144             72             27      228  33.9   
96            2       92             62             28        0  31.6   
97            1       71             48             18       76  20.4   
98            6       93             50             30       64  28.7   
99            1      122             90             51      220  49.7   

    DiabetesPedigreeFunction  Age Outcome  
0                      0.627   50       1  
1                      0.351   31       0  
2                      0.672   32       1  
3                      0.167   21       0  
4                      2.288   33       1  
..                       ...  ...     ...  
95                     0.255   40       0  
96                     0.130   24       0  
97                     0.323   22       0  
98                     0.356   23       0  
99                     0.325   31       1  

[100 rows x 9 columns]

In [99]:
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].mean())

In [100]:
diabetes_data.head(100)

Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0             6    148.0           72.0      35.000000   79.799479  33.6   
1             1     85.0           66.0      29.000000   79.799479  26.6   
2             8    183.0           64.0      20.536458   79.799479  23.3   
3             1     89.0           66.0      23.000000   94.000000  28.1   
4             0    137.0           40.0      35.000000  168.000000  43.1   
..          ...      ...            ...            ...         ...   ...   
95            6    144.0           72.0      27.000000  228.000000  33.9   
96            2     92.0           62.0      28.000000   79.799479  31.6   
97            1     71.0           48.0      18.000000   76.000000  20.4   
98            6     93.0           50.0      30.000000   64.000000  28.7   
99            1    122.0           90.0      51.000000  220.000000  49.7   

    DiabetesPedigreeFunction  Age Outcome  
0                      0.627   50       1  
1                      0.351   31       0  
2                      0.672   32       1  
3                      0.167   21       0  
4                      2.288   33       1  
..                       ...  ...     ...  
95                     0.255   40       0  
96                     0.130   24       0  
97                     0.323   22       0  
98                     0.356   23       0  
99                     0.325   31       1  

[100 rows x 9 columns]